File: src/soft.jl
using LinearAlgebra
#using IterativeSolvers

# Input parameters
fname = length(ARGS)>0 ? ARGS[1] : "sphere-1.905-600.msh"
th = length(ARGS)>1 ? parse(Float32,ARGS[2]) : 0.0
f0 = length(ARGS)>2 ? parse(Float32,ARGS[3]) : 38e3
c0 = length(ARGS)>3 ? parse(Float32,ARGS[4]) : 1480.0
k = 2π * f0 / c0
d = [cos(π/180 * th), sin(π/180 * th), 0.0]

# Read mesh file
lines = readlines(fname)
n = parse(Int, lines[5])
v = [parse.(Float64, split(lines[5 + i])[2:4]) for i in 1:n]
m = parse(Int, lines[n + 8])
e = [parse.(Int, split(lines[n + 8 + i])[6:8]) for i in 1:m]

# Triangle centers (P0)
x = [sum([v[e[i][j]] / 3 for j in 1:3]) for i in 1:m]
# Three points from each triangle (P1)
# m *=3; x = [sum([v[e[i÷3+1][j+1]]*(i%3==j ? 4 : 1)/6 for j in 0:2]) for i in 0:m-1]

# Helmholtz single layer operator
S = zeros(ComplexF64, m, m)
for i in 1:m
    for j in 1:m
        r = norm(x[i] - x[j])
        S[i, j] = r != 0 ? exp(im * k * r) / r / (4π) : im * k / (4π)
    end
end

# Incident field
f = [exp(im * k * x[i]' * d) for i in 1:m]

# Surface solution
φ = lu(S) \ (-f)
#φ = gmres(S,-f)

# Far-field directions
r1 = [[cos(π/180 * i), sin(π/180 * i), 0.0] for i in 0:359]
psc = [sum([exp(-im * k * x[j]' * r) / (4π) * φ[j] for j in 1:m]) for r in r1]

# Print scattering pattern
for i in 1:360
    println(i - 1, '\t', abs(psc[i]))
end

File: src/soft.py
# plane wave scattering from soft target in salt water
# input: msh ascii 2.2 file
# output: 1-deg scattering pattern
# MM 26.7.2025            |
from numpy import pi, sqrt, exp, sin, cos, array, zeros
import sys

# input parameters
fname = 'sphere-1.905-600.msh' if len(sys.argv)<=1 else sys.argv[1]
th = 0 if len(sys.argv)<=2 else float(sys.argv[2])
f0 = 38e3 if len(sys.argv)<=3 else float(sys.argv[3])
c0 = 1480 if len(sys.argv)<=4 else float(sys.argv[4])
k = 2*pi*f0/c0; d = [cos(pi/180*th),sin(pi/180*th),0]

# reading nodes and elements from mesh file
with open(fname) as f: lines = f.readlines()
n = int(lines[4])
v = [list(map(float,lines[5+i].split()[1:4])) for i in range(n)]
m = int(lines[n+7])
e = [list(map(int,lines[n+8+i].split()[5:8])) for i in range(m)]

# space - centers of surface triangles
x = [sum([array(v[e[i][j]-1])/3 for j in range(3)]) for i in range(m)]

# Helmholtz single layer operator
S = zeros((m,m),dtype=complex);
for i in range(m):
    for j in range(m):
        r = sqrt(sum((x[i]-x[j])**2))
        S[i][j] = exp(1j*k*r)/r/(4*pi) if r!=0 else 1j*k/(4*pi)

# incident field
f = array([exp(1j*k*p@d) for p in x])

# surface solution
from scipy.linalg import lu_factor, lu_solve
phi = lu_solve(lu_factor(S), -f)

# far field solution
r1 = [[cos(pi/180*i),sin(pi/180*i),0] for i in range(360)]
S = [[exp(-1j*k*p@r)/(4*pi) for p in x] for r in r1]
psc = S @ phi;    
print("\n".join(["%d\t%g"%(i,abs(psc[i])) for i in range(len(psc))]))
